resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CharsetManager.java

```diff
@@ -440,45 +440,60 @@ public void projectPreferencesChanged(IProject project) {
-			if (newCharset != null)
-				resourcesPreferences.put(ResourcesPlugin.PREF_ENCODING, newCharset);
-			else
-				resourcesPreferences.remove(ResourcesPlugin.PREF_ENCODING);
-			try {
-				resourcesPreferences.flush();
-			} catch (BackingStoreException e) {
-				IProject project = workspace.getRoot().getProject(resourcePath.segment(0));
-				String message = Messages.resources_savingEncoding;
-				throw new ResourceException(IResourceStatus.FAILED_SETTING_CHARSET, project.getFullPath(), message, e);
-			}
-			return;
+		if (IPath.ROOT.equals(resourcePath)) {
+			setCharsetForRoot(resourcePath, newCharset);
+		} else if (workspace.getRoot().findMember(resourcePath) instanceof Resource resource) {
+			setCharsetForResource(resourcePath, newCharset, resource);
 		}
-		// for all other cases, we set a property in the corresponding project
-		IResource resource = workspace.getRoot().findMember(resourcePath);
-		if (resource != null) {
-			try {
-				// disable the listener so we don't react to changes made by ourselves
-				Preferences encodingSettings = getPreferences(resource.getProject(), true, resource.isDerived(IResource.CHECK_ANCESTORS));
-				if (newCharset == null || newCharset.trim().length() == 0)
-					encodingSettings.remove(getKeyFor(resourcePath));
-				else
-					encodingSettings.put(getKeyFor(resourcePath), newCharset);
-				flushPreferences(encodingSettings, true);
-				if (resource instanceof IProject) {
-					IProject project = (IProject) resource;
-					ValidateProjectEncoding.scheduleProjectValidation(workspace, project);
-				}
-			} catch (BackingStoreException e) {
-				IProject project = workspace.getRoot().getProject(resourcePath.segment(0));
-				String message = Messages.resources_savingEncoding;
-				throw new ResourceException(IResourceStatus.FAILED_SETTING_CHARSET, project.getFullPath(), message, e);
+	}
+
+	private void setCharsetForRoot(IPath resourcePath, String newCharset) throws ResourceException {
+		IEclipsePreferences resourcesPreferences = InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
+		if (newCharset != null) {
+			resourcesPreferences.put(ResourcesPlugin.PREF_ENCODING, newCharset);
+		} else {
+			resourcesPreferences.remove(ResourcesPlugin.PREF_ENCODING);
+		}
+
+		try {
+			resourcesPreferences.flush();
+		} catch (BackingStoreException e) {
+			setCharsetForHasFailed(resourcePath, e);
+		}
+	}
+
+	private void setCharsetForResource(IPath resourcePath, String newCharset, IResource resource)
+			throws ResourceException {
+		try {
+			setResourceEncodingSettings(resourcePath, newCharset, resource);
+			if (resource instanceof Project project) {
+				ValidateProjectEncoding.scheduleProjectValidation(workspace, project);
 			}
+		} catch (BackingStoreException e) {
+			setCharsetForHasFailed(resourcePath, e);
 		}
 	}
 
+	private void setResourceEncodingSettings(IPath resourcePath, String newCharset, IResource resource)
```

Comment:
I don't see much value in having these 8 lines extracted out here if the only consumer is a 3 lines method.

Reply:
The whole refactoring is a [functioncal decomposition](https://en.wikipedia.org/wiki/Functional_decomposition) of setCharsetFor, and so is this function. My goal is for somebody completely new to this unit to be able to read it "newspaper"-style without having to distill the essence of the code.

I am wondering, is there any harm to having this extraction?

Reply:
Not much harm, but I just don't buy blindly the argument that more methods is making things clearer and easier to understand. It requires jumps to read and understand things, so it cost cognitive complexity, it adds symbols to the .class (making it slightly bigger)... I know there are dogmas of "no method should be more than X lines", but I've grown too old and grumpy to believe that. Code can be expressive without being scattered.
But it's more for the sake of discussion, not a blocker against merging this PR.